lumicks/pylake/force_calibration/power_spectrum.py

```diff
@@ -156,3 +157,105 @@ def plot(self, **kwargs):
             plt.title(f"Blocked Power spectrum (N={self.num_points_per_block})")
         else:
             plt.title("Power spectrum")
+
+
+def identify_peaks(
+    power_spectrum: PowerSpectrum,
+    model_fun: callable,
+    return_indices: bool = False,
+    peak_cutoff: float = 20.0,
+    baseline: float = 1.0,
+):
+    """Identify peaks in the power spectrum, based on an exponential probability model with
+    rate parameter (lambda) = 1. This means that the power spectrum cannot be blocked or windowed.
+
+    Parameters
+    ----------
+    power_spectrum : PowerSpectrum
+        The power spectrum of the Brownian motion of the bead in a trap
+    model_fun : callable
+        A function of one argument, frequency, that gives the theoretical power spectrum. The
+        function is used to normalize the experimental power spectrum
+    return_indices: bool
+        Also return the array indices for the identified peaks (see below)
+    peak_cutoff: float
+        Indicates what value of the normalized spectrum is deemed abnormally high. Default is 20.0,
+        which corresponds to a chance of about 2 in 1E9 that a peak of that magnitude occurs
+        naturally in an exponential distribution with rate parameter = 1.0. The minimum is 15.0.
+    baseline: float
+        The baseline level a peak needs to drop down to. Lower means that exponentially more data is
+        considered to be part of the peak. The default is 1.0, and the range of baseline is [0.0,
+        peak_cutoff]. No fit or data smoothing is performed, the peak starts or ends at the first
+        data point that satisfies the criterion.
+
+    Returns
+    -------
+    frequency_ranges: list of tuples (f_start, f_stop)
+        f_start is the frequency where a peak starts, and f_end is the frequency where a peak ends.
+    frequency_indices: list of tuples (start_index, stop_index)
+        start_index is the index in the array power_spectrum.frequencies where a peak starts.
+        Similarly, stop_index is the index in the array where a peak stops. This list of tuples is
+        only returned when return_indices is True.
+
+    Raises
+    ------
+    ValueError
+        Raises a ValueError if the function is called on a PowerSpectrum object with blocking
+        applied. This function only works for PowerSpectrum objects without blocking or windowing.
+    """
+
+    def find_ranges(list_of_numbers):
+        """Place a range of consecutive numbers in a tuple as (start, stop), and return a list of
+        tuples"""
+        ranges = []
+        for _, chunk in groupby(enumerate(list_of_numbers), lambda x: x[1] - x[0]):
+            chunk = list(chunk)
+            ranges += [(chunk[0][1], chunk[-1][1])]
```

Comment:
Not a big deal, but in python it's pretty common to have an inclusive start and an exclusive end when you specify a range. So in this case, that would mean going one beyond the index you find for the second element of a tuple. That said, I'd also be fine with keeping this if you have a strong preference.

Reply:
I'm good with sticking to the general consensus, will adapt

Reply:
So, I'm confused: it seems that `PowerSpectrum.in_range()` actually _excludes_ the first item in the tuple, and _includes_ the second? There probably is a good reason why it's like that, but I cannot think of it. Should I tune my range finding function to this situation?

Reply:
I thought about this, but as the ranges are never used for slicing (with exclusive end), but always used for indexing, it's much more convenient to keep the function as is. Otherwise, I would do +1 in the embedded function and -1 in the host function, which I don't think makes the code more readable. What do you think?

Reply:
> So, I'm confused: it seems that `PowerSpectrum.in_range()` actually excludes the first item in the tuple, and includes the second? There probably is a good reason why it's like that, but I cannot think of it. Should I tune my range finding function to this situation?

Legacy reasons. It was like that in the original LDAT implementation :(
We should probably mention that more explicitly in the function (it being non-standard).

> I thought about this, but as the ranges are never used for slicing (with exclusive end), but always used for indexing, it's much more convenient to keep the function as is. Otherwise, I would do +1 in the embedded function and -1 in the host function, which I don't think makes the code more readable. What do you think?

Yeah, fair point. You also only use this thing locally, so it's fine.

Reply:
I have a real data set where there is a double peak. The first (and smallest) peak has some width to the left of it (lower frequencies), but not to the right. Therefore, if `peak_idx` is the index of the frequency bin containing the peak, the frequency exclusion range would be `(fmin, fmax) = (power_spectrum.frequency[peak_idx - N]), power_spectrum.frequency[peak_idx])`, where `N` is the number of bins that contain some of the peak to the left. 
See here:
![image](https://github.com/lumicks/pylake/assets/55446628/1d869b2d-1507-412a-b536-5d1cbfc69244)

With the current implementation of frequency exclusion ranges, which excludes everything up to `fmax`, the peak itself is actually never excluded. Is it acceptable to add an offset of `df`, where `df` is the spacing of the frequency bins? Typically, just returning the next higher bin (which is nearly the equivalent) would also work. However, this approach fails when the peak is at the right edge of the spectrum, as that bin does not exist. What do you think?

Reply:
Yeah, I think if you want inclusive end (which you do in this case) you're gonna have to add the bin width to the upper limit. I think that's fine.